### PR TITLE
Leaves - Morgan

### PR DESCRIPTION
**Pull request by morganschuler**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is a class that can add records to a database in rails. It is the one that does the CRUD operations. 
Describe in your own words what the Controller is doing in Rails | The controller is the one that is in the middle of everything! It handles requests from the router, communicates with the data/database in the model, and communicates with the view about what will appear.
Describe in your own words what the View is doing in Rails | The view receives data from the controller and sends it back to the browser for a display. 
Describe an edge-case controller test you wrote | I wrote an edge case that will redirect to the root_path in the event of an invalid id. 
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params save a lot of time and typing space so you don't have to go back and retrieve everything from the hash every time but it is also a lot safer, as params have to be permitted. Therefore not everyone can mess with them.
How are Rails migrations related to Rails models? | Migrations are related to models because when you add a new model to a Rails application, it generates a migration to create the corresponding table for you. 
Describe one area of Rails that are still unclear on | I still feel a little shaky about migrations and how to fix them in the event that I miss one up. It hasn't happened yet but I'm a little nervous for when that day arrives.

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    end
+  end
+  
+  def update
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      redirect_to root_path
+      return
+    end
+    if @task.update(task_params)
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:53]: This is not working. Investigate by looking at your `task_params`.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -71,56 +64,126 @@
...
+    }
     it "can update an existing task" do
-      # Your code here
+      expect {
+        patch task_path(@new_task.id), params: updated_task_form_data
+      }.wont_change "Task.count"
+      
+      @new_task.reload
+      
+      expect(@new_task.name).must_equal "Laundry"
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:124]: It's not clear to me how this is testing the update method as `@new_task` had the name 'Laundry' from the start.

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -71,56 +64,126 @@
...
-    # Your tests go here
-
-  end
-
-  # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
-  end
-end
+    it "can destroy a task" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:137]: You need to test `destroy` and `toggle_complete` and `toggle_incomplete` for valid and **invalid** task ids.

**Comment by beccaelenzil** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | check
Answered comprehension questions | check
Successfully handles: Index, Show | check
Index & Show tests pass |check
Successfully handles: New, Create | check
New & Create tests pass |check
Successfully handles: Edit, Update | entries in the edit form are not saved the the database
Tests for Edit & Update test valid & invalid task ids | see comment
Successfully handles: Destroy, Task Complete | check
Tests for Destroy & Task Complete include tests for valid and invalid task ids | see comment
Routes follow RESTful conventions | check
Uses named routes (like `_path`) | check
**Overall** | You hit a number of learning goals including using partial, view_helpers, and prefixes. You are clearly on your way to using rails conventions fluently. For your tests, you need to make sure to always test for valid and invalid input. In this case, invalid input is an invalid id. The update action needs a small change to fully function. Compare `create` to `update` to see if you can figure out what is going on.
